identity-did/src/document/core_document.rs

```diff
@@ -218,48 +251,109 @@ impl<T, U, V> CoreDocument<T, U, V> {
     &mut self.properties
   }
 
-  /// Maps `CoreDocument<T>` to `CoreDocument<U>` by applying a function to the custom
-  /// properties.
-  pub fn map<A, F>(self, f: F) -> CoreDocument<A, U, V>
+  /// Maps `CoreDocument<D,T>` to `CoreDocument<C,U>` by applying a function `f` to all [`DID`] fields
+  /// and another function `g` to the custom properties.
+  // TODO: remove this? Unused but useful due to lack of From specialization.
```

Comment:
I am concerned about the amount of responsibilities the function `f` has here. Perhaps the [Fold pattern](https://rust-unofficial.github.io/patterns/patterns/creational/fold.html) might be a better fit, but that introduces more code and complexity to the code base. Hence I think we should remove this for now and reconsider what to do if someone requests such functionality in the future.

Reply:
The `map` and `try_map` functions were present before and they are technically needed to convert between `CoreDocument` instances with different generics. I would not use fold since `f` only has one responsibility which is converting DIDs from one type to another, e.g. `IotaDID` to `CoreDID`.

Reply:
To ble clear I like the map implementations elsewhere, it is just this one I am concerned about.

Reply:
Should we revert back to the previous behaviour where it mapped the generic properties alone or remove it completely? 

Previously the properties were more important because that was the intended mechanism for extending documents with e.g. metadata and proof fields. It's less important now after the metadata refactor which went with a composition strategy.

Reply:
I would say either keep it as it is now or remove it completely. 

> since f only has one responsibility which is converting DIDs from one type to another

If this is the only or at least most common use case then I would just keep it as it is.

Reply:
> If this is the only or at least most common use case then I would just keep it as it is.

```rust
    C: DID + KeyComparable,
    F: FnMut(D) -> C,
```

It's the only possible use case, unless someone implements `DID` for an arbitrary struct. 😅

Reply:
This does seem useful to me, and since it's already implemented, I'd also be fine with keeping it.